### PR TITLE
test: Add support for querying shadow DOM

### PR DESCRIPTION
**Pull request by martinpitt**
Add a `querySelectorAll()` variant which can transparently pierce and traverse shadow DOMs without having to know their structure (i.e. doing selectors piece-wise and iterating on their `.shadowRoot` attribute).

As this is more expensive than standard querySelector()`, only  use it if the page actually contains any shadow root.
    
This makes it a lot more convenient to write tests for pages which use web components, like https://patternflyelements.org.

----

https://github.com/cockpit-project/starter-kit/pull/1028 makes use of that.

TODO:
  - [x] This makes queries much slower. Provide an opt-in.

### CODE REVIEW

**Comment by martinpitt** [comment]
Apparently this [does not work in Firefox](https://artifacts.dev.testing-farm.io/865766c1-8ace-4911-9bcc-e0b5e7f72265/), needs some more work. Also, this isn't urgent at all.

**Comment by martinpitt** [comment]
The previous commit f11e9e8c3aedb119f8e9e38d889bc1e00956b16b always used this "deep" querySelector, and it worked. But I figure it'll be slower, so I reworked this to detect the presence of shadow DOMs, and only use the deep one if there are any.

I adjusted https://github.com/cockpit-project/starter-kit/pull/1028 to pull in this new version.

File: test/common/test-functions.js [hunk 1]
```diff
@@ -4,12 +4,38 @@
  * These are routines used by our testing code.
  */
 
+/* Detect if we have any shadow DOM */
+window.__haveShadowDom = function() {
+    if (window.__haveShadowDomResult === undefined)
+        window.__haveShadowDomResult = Array.from(document.querySelectorAll('*')).filter(el => el.shadowRoot).length > 0;
```
Comment (jelly) [inline test/common/test-functions.js:10]: This seems pretty expensive? Likely it is more performance to write:
```
for (int x = 0 ; i < querySelector('*').length; i++ )
    if (el.shadowRoot)
         return true;
```

Alternatively:

```
let has_shadow_dom = false
document.querySelectorAll('*').forEach(element => {if (element.shadowRoot) { has_shadow_dom = true ;break; }})
```

Alternatively `document.body.contains()` would I suppose "work" but requires you to run it every time.

Comment (martinpitt) [reply test/common/test-functions.js]: Right, I rewrote it using find(). Note that this isn't super-critical, it just runs once for every frame.

I don't see how .contains() could help? that just checks if a node is a child of another node, but I don't have any child node here.

**Review by jelly — APPROVED** [review]

**Pull request merged**